search/retrieval.cpp

```diff
@@ -53,17 +55,57 @@ void CoverRect(m2::RectD const & rect, int scale, covering::IntervalsT & result)
   result.insert(result.end(), intervals.begin(), intervals.end());
 }
 
+bool MatchFeatureByName(FeatureType const & ft, SearchQueryParams const & params)
+{
+  bool matched = false;
+  auto const prefix =
+      params.m_prefixTokens.empty() ? SearchQueryParams::TString() : params.m_prefixTokens.front();
+  ft.ForEachName([&](int8_t /*lang*/, string const & name) -> bool
+                 {
+                   auto const nsName = search::NormalizeAndSimplifyString(name);
+                   if (!prefix.empty() && strings::StartsWith(nsName, prefix))
+                   {
+                     matched = true;
+                     return false;
+                   }
+                   for (auto const & synonyms : params.m_tokens)
+                   {
+                     for (auto const & token : synonyms)
+                     {
+                       if (token == nsName)
+                       {
+                         matched = true;
+                         return false;
+                       }
+                     }
+                   }
+                   return true;  // To iterate all names.
+                 });
+  return matched;
+}
+
 // Retrieves from the search index corresponding to |value| all
 // features matching to |params|.
 template <typename TValue>
 unique_ptr<coding::CompressedBitVector> RetrieveAddressFeaturesImpl(
-    MwmValue & value, my::Cancellable const & cancellable, SearchQueryParams const & params)
+    MwmSet::MwmId const & id, MwmValue & value, my::Cancellable const & cancellable,
+    SearchQueryParams const & params)
 {
   serial::CodingParams codingParams(trie::GetCodingParams(value.GetHeader().GetDefCodingParams()));
   ModelReaderPtr searchReader = value.m_cont.GetReader(SEARCH_INDEX_FILE_TAG);
 
-  auto emptyFilter = [](uint32_t /* featureId */)
+  // Exclude from search all deleted/modified features and match all edited/created features separately.
+  using osm::Editor;
+  Editor & editor = Editor::Instance();
+  auto const deleted = editor.GetFeaturesByStatus(id, Editor::FeatureStatus::Deleted);
+  // Modified features shoud be re-matched again.
+  auto modified = editor.GetFeaturesByStatus(id, Editor::FeatureStatus::Modified);
+  auto const filter = [&](uint32_t featureIndex) -> bool
   {
+    if (find(deleted.begin(), deleted.end(), featureIndex) != deleted.end())
```

Comment:
We did not make any measurements yet. But it's not feasible to manually edit thousands features from mobile device between mwm update cycle.

Users are allowed to edit features only. All local edits are supposed to be cleared after successful upload to OSM and downloading next MWM version. Bad performance can always be improved in the next release by refactoring editor's core, and core refactoring is in our plans.